# Problem 1 - Word counts (40 points)

## Part A. Characters in The Hound of the Baskervilles

Use the text available at https://s3.amazonaws.com/2018-dmfa/project-1/hound.txt for this part.

How many times are each of the following characters mentioned by name in the text of The Hound of the Baskervilles?

* Holmes, Watson, Barrymore, Mortimer

##### Get the text file

In [1]:
!wget https://s3.amazonaws.com/2018-dmfa/project-1/hound.txt

--2018-09-24 01:55:11--  https://s3.amazonaws.com/2018-dmfa/project-1/hound.txt
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.18.163
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.18.163|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 345659 (338K) [text/plain]
Saving to: ‘hound.txt’

hound.txt           100%[===================>] 337.56K  --.-KB/s    in 0.006s  

2018-09-24 01:55:11 (57.9 MB/s) - ‘hound.txt’ saved [345659/345659]



##### Take a peek

In [3]:
!head hound.txt

Project Gutenberg’s The Hound of the Baskervilles, by A. Conan Doyle

This eBook is for the use of anyone anywhere at no cost and with
almost no restrictions whatsoever.  You may copy it, give it away or
re-use it under the terms of the Project Gutenberg License included
with this eBook or online at www.gutenberg.org


Title: The Hound of the Baskervilles



##### We exclude cases like "Barrymores" with an "s", since it's not mentioned as the character's name

In [4]:
!grep -wc -i Holmes hound.txt
!grep -wc -i Watson hound.txt
!grep -wc -i Barrymore hound.txt
!grep -wc -i Mortimer hound.txt

192
113
72
89


- Holmes is mentioned by name 192 times in the text;
- Watson is mentioned by name 113 times in the text;
- Barrymore is mentioned by name 72 times in the text;
- Mortimer is mentioned by name 89 times in the text.

## Part B. Characters in Hamlet

Use the text available at https://s3.amazonaws.com/2018-dmfa/project-1/hamlet.txt for this part.

How many times do each of the characters Hamlet, Polonius, Ophelia, and Horatio have speaking lines in Hamlet? Keep in mind that this is the text of a play.

##### Get the text file

In [5]:
!wget https://s3.amazonaws.com/2018-dmfa/project-1/hamlet.txt

--2018-09-24 01:55:41--  https://s3.amazonaws.com/2018-dmfa/project-1/hamlet.txt
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.232.37
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.232.37|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 204885 (200K) [text/plain]
Saving to: ‘hamlet.txt’

hamlet.txt          100%[===================>] 200.08K  --.-KB/s    in 0.006s  

2018-09-24 01:55:41 (33.2 MB/s) - ‘hamlet.txt’ saved [204885/204885]



##### When character speaks, there are 2 cases:
- case 1: the name is capitalised and followed by a dot
- case 2: 2 charactor speak together, for example, "HORATIO and MARCELLUS."

In [6]:
!cat hamlet.txt | grep -e ^HAMLET'[[:space:]]' -e ^HAMLET'[.]' | wc -l
!cat hamlet.txt | grep -e ^POLONIUS'[[:space:]]' -e ^POLONIUS'[.]' | wc -l
!cat hamlet.txt | grep -e ^OPHELIA'[[:space:]]' -e ^OPHELIA'[.]' | wc -l
!cat hamlet.txt | grep -e ^HORATIO'[[:space:]]' -e ^HORATIO'[.]' | wc -l

358
86
58
110


- Hamlet has 358 speaking lines in Hamlet;
- Polonius has 86 speaking lines in Hamlet;
- Ophelia has 58 speaking lines in Hamlet;
- Horatio has 110 speaking lines in Hamlet.

# Problem 2 - Capital Bikeshare (40 points)

Use the data available at https://s3.amazonaws.com/2018-dmfa/project-1/2018-capitalbikeshare-tripdata.zip for this problem.

## Part A (20 points)

1. Unzip the data and combine the two CSV files using csvstack.

2. Which 10 Capital Bikeshare stations were the most popular departing stations in July and August 2018? Which 10 stations were the most popular destination stations in July and August 2018? For both answers (and for the following questions as well), provide the full name, not just the station number.

##### Get the zip file

In [7]:
!wget https://s3.amazonaws.com/2018-dmfa/project-1/2018-capitalbikeshare-tripdata.zip

--2018-09-24 01:55:59--  https://s3.amazonaws.com/2018-dmfa/project-1/2018-capitalbikeshare-tripdata.zip
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.129.189
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.129.189|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18365732 (18M) [application/zip]
Saving to: ‘2018-capitalbikeshare-tripdata.zip’

2018-capitalbikesha 100%[===================>]  17.51M   106MB/s    in 0.2s    

2018-09-24 01:55:59 (106 MB/s) - ‘2018-capitalbikeshare-tripdata.zip’ saved [18365732/18365732]



##### Unzip the file

In [8]:
!unzip 2018-capitalbikeshare-tripdata.zip

Archive:  2018-capitalbikeshare-tripdata.zip
  inflating: 201807-capitalbikeshare-tripdata.csv  
  inflating: 201808-capitalbikeshare-tripdata.csv  


##### Combine two csv using csvstack

In [9]:
!csvstack 201807-capitalbikeshare-tripdata.csv 201808-capitalbikeshare-tripdata.csv > combined.csv

##### Find out the most popular 10 departing stations in July and August 2018

In [10]:
!csvcut combined.csv -c4,5 | sort | uniq -c | sort -rn | head -10 | csvlook

|   15474 31258 | Lincoln Memorial                                      |
| ------------- | ----------------------------------------------------- |
|   13082 31623 | Columbus Circle / Union Station                       |
|   11999 31247 | Jefferson Dr & 14th St SW                             |
|   11671 31289 | Henry Bacon Dr & Lincoln Memorial Circle NW           |
|   11650 31288 | 4th St & Madison Dr NW                                |
|   10357 31248 | Smithsonian-National Mall / Jefferson Dr & 12th St SW |
|    9554 31249 | Jefferson Memorial                                    |
|    8620 31290 | 17th St & Independence Ave SW                         |
|    8583 31200 | Massachusetts Ave & Dupont Circle NW                  |
|    8569 31201 | 15th & P St NW                                        |


##### Find out the most popular 10 destination stations in July and August 2018

In [11]:
!csvcut combined.csv -c6,7 | sort | uniq -c | sort -rn | head -10 | csvlook

|   15642 31258 | Lincoln Memorial                                      |
| ------------- | ----------------------------------------------------- |
|   13635 31623 | Columbus Circle / Union Station                       |
|   12135 31247 | Jefferson Dr & 14th St SW                             |
|   11722 31289 | Henry Bacon Dr & Lincoln Memorial Circle NW           |
|   11555 31288 | 4th St & Madison Dr NW                                |
|   10693 31248 | Smithsonian-National Mall / Jefferson Dr & 12th St SW |
|    9866 31249 | Jefferson Memorial                                    |
|    9141 31200 | Massachusetts Ave & Dupont Circle NW                  |
|    8884 31201 | 15th & P St NW                                        |
|    8640 31290 | 17th St & Independence Ave SW                         |


## Part B (20 points)

1. For the most popular departure station, which 10 bikes were used most in trips departing from there?

2. Which 10 bikes were used most in trips ending at the most popular destination station?

##### Find the 10 bikes which were used most in trips departing from the most popular departure station, No.31258 Lincoln Memorial.

In [12]:
!csvcut combined.csv -c4,5,8 | csvgrep -c1 -r "31258" | sort | uniq -c | sort -rn | head -10 | csvlook

sort: write failed: 'standard output': Broken pipe
sort: write error
|      18 31258 | Lincoln Memorial | W01311 |
| ------------- | ---------------- | ------ |
|      17 31258 | Lincoln Memorial | W22553 |
|      16 31258 | Lincoln Memorial | W00919 |
|      15 31258 | Lincoln Memorial | W23074 |
|      15 31258 | Lincoln Memorial | W22567 |
|      15 31258 | Lincoln Memorial | W21882 |
|      15 31258 | Lincoln Memorial | W20527 |
|      14 31258 | Lincoln Memorial | W23311 |
|      14 31258 | Lincoln Memorial | W23003 |
|      14 31258 | Lincoln Memorial | W22369 |


##### Find the 10 bikes which were used most in trips ending at the most popular destination station, No.31258 Lincoln Memorial.

In [13]:
!csvcut combined.csv -c6,7,8 | csvgrep -c1 -r "31258" | sort | uniq -c | sort -rn | head -10 | csvlook

sort: write failed: 'standard output': Broken pipe
sort: write error
|      17 31258 | Lincoln Memorial | W22553 |
| ------------- | ---------------- | ------ |
|      17 31258 | Lincoln Memorial | W01311 |
|      16 31258 | Lincoln Memorial | W21882 |
|      16 31258 | Lincoln Memorial | W20527 |
|      16 31258 | Lincoln Memorial | W00919 |
|      15 31258 | Lincoln Memorial | W22567 |
|      15 31258 | Lincoln Memorial | W22369 |
|      14 31258 | Lincoln Memorial | W23311 |
|      14 31258 | Lincoln Memorial | W23074 |
|      14 31258 | Lincoln Memorial | W23003 |


# Problem 3 - Filters (20 points)

In class lectures, previous exercises, and Problems 1 and 2 above, you use Unix commands like grep and tr as filters, changing the text lines streaming through the pipeline. In this problem, write small Python or R programs that act as a filter in the same way, where each program serves one filtering purpose. Name your new filter programs clearly.

For this problem, use the Python filter template shown in class and available at https://s3.amazonaws.com/2018-dmfa/project-1/simplefilter.py 

or the R filter template available at https://s3.amazonaws.com/2018-dmfa/project-1/simplefilter.R as the basis of your own filters.

## Part A (10 points)

1. Demonstrate a pipeline that performs a count of the top ten most common words in The Hound of the Baskervilles. This may be exactly the same pipeline we have used before.

2. Write your own lowercase filter that replaces tr [:upper:] [:lower:] to transform text into lower case.

3. With your new filter, repeat the original pipeline, and substitute your new filters as appropriate. You should obtain the same results as in 3.A.1 above.

##### Get Python file

In [14]:
!wget https://s3.amazonaws.com/2018-dmfa/project-1/simplefilter.py

--2018-09-24 01:57:15--  https://s3.amazonaws.com/2018-dmfa/project-1/simplefilter.py
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.20.245
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.20.245|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 254 [text/x-python-script]
Saving to: ‘simplefilter.py’

simplefilter.py     100%[===================>]     254  --.-KB/s    in 0s      

2018-09-24 01:57:15 (9.20 MB/s) - ‘simplefilter.py’ saved [254/254]



##### Get R file

In [15]:
!wget https://s3.amazonaws.com/2018-dmfa/project-1/simplefilter.R

--2018-09-24 01:57:18--  https://s3.amazonaws.com/2018-dmfa/project-1/simplefilter.R
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.166.5
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.166.5|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 244 [binary/octet-stream]
Saving to: ‘simplefilter.R’

simplefilter.R      100%[===================>]     244  --.-KB/s    in 0s      

2018-09-24 01:57:19 (12.2 MB/s) - ‘simplefilter.R’ saved [244/244]



##### Count of the top ten most common words in The Hound of the Baskervilles.

In [16]:
!cat hound.txt | tr -sc '[:alpha:]' '[\n*]' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort -rn | head -10

   3515 the
   1716 of
   1696 and
   1501 i
   1488 to
   1369 a
   1159 that
   1025 it
    968 in
    922 he
sort: write failed: 'standard output': Broken pipe
sort: write error


### Solve the question 2 and 3 using Python codes

#### Write your own lowercase filter that replaces tr [:upper:] [:lower:] to transform text into lower case

##### Modify the Python file as below and rename it as lower.py

##### Change permissions

In [17]:
!chmod +x lower.py

##### Have a try to apply lowercase filter to hound.txt

In [18]:
!head hound.txt | ./lower.py

project gutenberg’s the hound of the baskervilles, by a. conan doyle

this ebook is for the use of anyone anywhere at no cost and with
almost no restrictions whatsoever.  you may copy it, give it away or
re-use it under the terms of the project gutenberg license included
with this ebook or online at www.gutenberg.org


title: the hound of the baskervilles



##### With your new filter, repeat the original pipeline, and substitute your new filters as appropriate

In [19]:
!cat hound.txt | tr -sc '[:alpha:]' '[\n*]' | ./lower.py | sort | uniq -c | sort -rn | head -10

   3515 the
   1716 of
   1696 and
   1501 i
   1488 to
   1369 a
   1159 that
   1025 it
    968 in
    922 he
sort: write failed: 'standard output': Broken pipe
sort: write error


### Solve the question 2 and 3 using R codes

#### Write your own lowercase filter that replaces tr [:upper:] [:lower:] to transform text into lower case

##### Modify the R file as below and rename it as lower.R

##### Change permissions

In [20]:
!chmod +x lower.R

##### Have a try to apply lowercase filter to hound.txt

In [21]:
!head hound.txt | ./lower.R

project gutenberg’s the hound of the baskervilles, by a. conan doyle

this ebook is for the use of anyone anywhere at no cost and with
almost no restrictions whatsoever.  you may copy it, give it away or
re-use it under the terms of the project gutenberg license included
with this ebook or online at www.gutenberg.org


title: the hound of the baskervilles



##### With your new filter, repeat the original pipeline, and substitute your new filters as appropriate

In [22]:
!cat hound.txt | tr -sc '[:alpha:]' '[\n*]' | ./lower.R | sort | uniq -c | sort -rn | head -10

   3515 the
   1716 of
   1696 and
   1501 i
   1488 to
   1369 a
   1159 that
   1025 it
    968 in
    922 he
sort: write failed: 'standard output': Broken pipe
sort: write error


## Part B (10 points)

1. Write a filter that removes at least ten common words of English text, commonly known as “stop words”. Sources of English stop word lists are readily available online, or you may generate your own list from the text. Document your process for obtaining or determining this list.

2. Add your stop word filter to a word count pipeline and show the most common 25 words in The Hound of the Baskervilles when stop words are removed. You may re-use your filter from Part A if you wish, although this is not required for full credit.

##### We find a txt file link from the Wikipedia page of 'Stop words'. Get stop-word-list.txt file from the link below.

In [23]:
!wget http://xpo6.com/wp-content/uploads/2015/01/stop-word-list.txt

--2018-09-24 01:57:56--  http://xpo6.com/wp-content/uploads/2015/01/stop-word-list.txt
Resolving xpo6.com (xpo6.com)... 104.31.84.205, 104.31.85.205
Connecting to xpo6.com (xpo6.com)|104.31.84.205|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2231 (2.2K) [text/plain]
Saving to: ‘stop-word-list.txt’

stop-word-list.txt  100%[===================>]   2.18K  --.-KB/s    in 0s      

2018-09-24 01:57:57 (11.3 MB/s) - ‘stop-word-list.txt’ saved [2231/2231]



### Solve the problem using Python codes

##### Modify the Python file as below and rename it as stop_words_remove.py

##### Change permissions

In [5]:
!chmod +x stop_words_remove.py

##### Explain the pipeline: show all contents | spilit into one words each line | run the lowercase filter | run the removing stop words filter | get rid of the lines of 'None' | sort and count words

##### Because in the Python file we replace the stop words as 'None', in our pipeline we add a filter to get rid of the lines of 'None'.

In [26]:
!cat hound.txt | tr -sc '[:alpha:]' '[\n*]' | ./lower.py | ./stop_words_remove.py | grep -v 'None' | sort | uniq -c | sort -rn | head -25

    350 sir
    247 s
    240 said
    213 man
    193 holmes
    163 moor
    156 henry
    119 know
    114 watson
    114 baskerville
    113 did
    109 dr
     94 charles
     93 stapleton
     93 gutenberg
     90 think
     90 mortimer
     89 come
     87 project
     86 time
     77 face
     76 hound
     74 came
     73 night
     73 house
sort: write failed: 'standard output': Broken pipe
sort: write error


### Solve the problem using R code

##### Modify the R file as below and rename it as stop_words_remove.R

##### Change permissions

In [26]:
!chmod +x stop_words_remove.R

##### Explain the pipeline: show all contents | spilit into one words each line | run the lowercase filter | run the removing stop words filter | only get the lines containing word | sort and count words

##### Because in the R file we replace the stop words as 'NULL', in our pipeline we add a filter to get only the lines containing word rather than 'NULL'.

In [27]:
!cat hound.txt | tr -sc '[:alpha:]' '[\n*]' | ./lower.R | ./stop_words_remove.R | grep '[[:print:]]' | sort | uniq -c | sort -rn | head -25

    350 sir
    247 s
    240 said
    213 man
    193 holmes
    163 moor
    156 henry
    119 know
    114 watson
    114 baskerville
    113 did
    109 dr
     94 charles
     93 stapleton
     93 gutenberg
     90 think
     90 mortimer
     89 come
     87 project
     86 time
     77 face
     76 hound
     74 came
     73 night
     73 house
sort: write failed: 'standard output': Broken pipe
sort: write error
